codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -359,16 +359,15 @@ public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, lon
     @Override
     public ChannelFuture writeGoAway(ChannelHandlerContext ctx, int lastStreamId, long errorCode, ByteBuf debugData,
             ChannelPromise promise) {
-        Http2Connection connection = connection();
-        if (connection.isGoAway()) {
+        if (connection().goAwayReceived()) {
             debugData.release();
-            return ctx.newSucceededFuture();
+            promise.setSuccess();
+            return promise;
         }
-
         ChannelFuture future = frameWriter().writeGoAway(ctx, lastStreamId, errorCode, debugData, promise);
         ctx.flush();
 
-        connection.goAwaySent(lastStreamId);
+        connection().goAwaySent(lastStreamId);
```

Comment:
Yup ... +1 to calling `goAwaySent` before the write so that we can catch the exception.

Also +1 to making `goAwaySent()` throw `Http2Exception` to make this explicit.